drivers/i3c/i3c_stm32.c

```diff
@@ -0,0 +1,2129 @@
+	struct i3c_stm32_msg *curr_msg = &data->curr_msg;
+
+	data->fifo_len = curr_msg->num_msgs * sizeof(uint32_t);
+	data->control_fifo = k_heap_alloc(&stm32_i3c_fifo_heap, data->fifo_len, K_FOREVER);
+	data->status_fifo = k_heap_alloc(&stm32_i3c_fifo_heap, data->fifo_len, K_FOREVER);
+	int ret;
+
+	/* Prepare all control words for all messages on the transfer */
+	for (size_t i = 0; i < curr_msg->num_msgs; i++) {
+		WRITE_REG(data->control_fifo[i],
+			  ((curr_msg->target_addr << I3C_CR_ADD_Pos) |
+			   i3c_stm32_curr_msg_control_get_len(dev) |
+			   i3c_stm32_curr_msg_control_get_dir(dev) | curr_msg->msg_type |
+			   i3c_stm32_curr_msg_control_get_end(dev)) &
+				  (I3C_CR_ADD | I3C_CR_DCNT | I3C_CR_RNW | I3C_CR_MTYPE |
+				   I3C_CR_MEND));
+
+		i3c_stm32_curr_msg_control_next(dev);
+	}
+
+	/* Configure DMA for the first message only, DMA callback will take care of the rest */
+	uint8_t *buf = NULL;
+	size_t *offset = 0;
+	uint32_t len = 0;
+
+	i3c_stm32_curr_msg_xfer_get_buf(dev, &buf, &len, &offset);
+
+	ret = i3c_stm32_dma_msg_config(dev, (uint32_t)buf, len);
+	if (ret != 0) {
+		return ret;
+	}
+
+	ret = i3c_stm32_dma_msg_control_fifo_config(dev);
+	if (ret != 0) {
+		return ret;
+	}
+
+	ret = i3c_stm32_dma_msg_status_fifo_config(dev);
+	if (ret != 0) {
+		return ret;
+	}
+
+	i3c_stm32_prepare_dma_requests(dev);
+#endif
+
+	/* Begin transmission */
+	LL_I3C_RequestTransfer(i3c);
+
+	/* Wait for whole transfer to complete */
+	if (k_sem_take(&data->device_sync_sem, STM32_I3C_TRANSFER_TIMEOUT) != 0) {
+		return -ETIMEDOUT;
+	}
+
+	if (data->msg_state == STM32_I3C_MSG_ERR) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+/* Handles the controller private read/write transfers */
+static int i3c_stm32_i3c_transfer(const struct device *dev, struct i3c_device_desc *target,
+				  struct i3c_msg *msgs, uint8_t num_msgs)
+{
+	struct i3c_stm32_data *data = dev->data;
+	int ret;
+
+	/* Verify all messages */
+	for (size_t i = 0; i < num_msgs; i++) {
+		if (msgs[i].buf == NULL) {
+			return -EINVAL;
+		}
+		if ((msgs[i].flags & I3C_MSG_HDR) && (msgs[i].hdr_mode != 0)) {
+			return -ENOTSUP;
+		}
+	}
+
+	k_mutex_lock(&data->bus_mutex, K_FOREVER);
+	ret = i3c_stm32_curr_msg_init(dev, msgs, NULL, num_msgs, target->dynamic_addr);
+	if (ret != 0) {
+		i3c_stm32_clear_err(dev, false);
+		LOG_ERR("Failed to initialize transfer messages, err=%d", ret);
+		return ret;
+	}
+
+	ret = i3c_stm32_transfer_begin(dev);
+	if (ret != 0) {
+		i3c_stm32_clear_err(dev, false);
+		LOG_ERR("Failed to transfer messages, err=%d", ret);
+		return ret;
+	}
+
+#ifdef CONFIG_I3C_STM32_DMA
+	/* Fill the num_xfer for each message from the status FIFO */
+	for (size_t i = 0; i < num_msgs; i++) {
+		msgs[i].num_xfer = READ_BIT(data->status_fifo[i], I3C_SR_XDCNT);
+	}
+
+	k_heap_free(&stm32_i3c_fifo_heap, data->control_fifo);
+	k_heap_free(&stm32_i3c_fifo_heap, data->status_fifo);
+
+	i3c_stm32_end_dma_requests(dev);
+#endif
+
+	k_mutex_unlock(&data->bus_mutex);
+
+	return 0;
+}
```

Comment:
There is the `I3C_MSG_NBCH` within `flags` that should be used to disable/enable the arbitration 0x7e flag